### ISSUE TITLE
[Implementation Error] nltk.metrics.windowdiff and nltk.metrics.pk

### ISSUE THREAD

**Issue by Coldog2333** [nltk/nltk#3020]
## API that has an error
`nltk.metrics.windowdiff`

## Wrong Codes
Line 92
```
return wd / (len(seg1) - k + 1.0)
```

## Corrections
Line 92
Should be revised as
```
return wd / (len(seg1) - k)
```

## Detailed Descriptions
Hello, thanks for your contribution to this useful toolkit.

Recently, I found a mistake in the implementation of `nltk.metrics.windowdiff`.

According to the formula on the 8th page of the paper _'Pevzner, L., & Hearst, M. A. (2002). A critique and improvement of an evaluation metric for text segmentation. Computational Linguistics, 28(1), 19-36.'_, the denominator of WindowDiff is `N-k`, where `k` is the window size. However, in the implementation of `nltk.metrics.windowdiff`, `N-k+1` was used, which will slightly affect the result. 

It should be better to correct it ASAP.

**Comment by Coldog2333** [comment]
What's more, API: `nltk.metrics.pk` has the same problem.

**Comment by tomaarsen** [comment]
#345 has more information on the changes that introduced `k + 1` for `windowdiff`. 
Furthermore, #353 has information caused the `k + 1` for `pk`. 

It does seem like the intention was to correspond with these formulas.

**Comment by tomaarsen** [comment]
I've had a look at #345, the current code, and https://aclanthology.org/J02-1002.pdf (p26), and it makes me suspect that your correction will not work. Note that in the formula, although the denominator is `N-k`, the loop is also from `1` to `N-k`, while our loop is from `0` (inclusive) to `N-k + 1` (exclusive). 
![image](https://user-images.githubusercontent.com/37621491/178726844-bc14d15c-a5a8-48c0-8cdb-0071f1f43576.png)
Note also that in this formula, `ref_i` with `i = 1` refers to the first element in the reference.

To further back up why I think your correction will not work: see the following example with your correction in place:
```python
# With NLTK's WindowDiff Corrected as per #3020
from nltk.metrics import windowdiff

s4 = "000000000000"
s5 = "111111111111"
print(windowdiff(s4, s5, 3))
```
This produces
```python
1.1111111111111112
```
Without the correction, it simply becomes `1`.

So, it simply does one more iteration than expected, meaning that the denominator must be one larger too. Note also that the number of iterations does seem to be correct - it considers all the windows that it should. If I modify the WindowDiff implementation to also print which windows it is operating on, like so:
```python
    wd = 0
    for i in range(len(seg1) - k + 1):
        print(seg1[i : i + k], seg2[i : i + k])
        ndiff = abs(seg1[i : i + k].count(boundary) - seg2[i : i + k].count(boundary))
        if weighted:
            wd += ndiff
        else:
            wd += min(1, ndiff)
    return wd / (len(seg1) - k + 1.0)
```
And then run the following script:
```python
# With NLTK's existing WindowDiff implementation, with a print added
from nltk.metrics import windowdiff

s1 = "000100000010"
s2 = "000010000100"
print(windowdiff(s1, s2, 3))
```
Then the output becomes this:
```python
000 000
001 000
010 001
100 010
000 100
000 000
000 000
000 001
001 010
010 100
0.3
```
And after a bit of careful comparison, we can see that the algorithm correctly starts at the very first full window, and ends at the very last full window.

In short, I think the existing implementation is correct, despite seeming a bit weird when visually compared to the formula.

**Comment by Coldog2333** [comment]
Thank you for your detailed explanation!

Yes, you are right.

Based on this, I think the implementations of Windowdiff and Pk in `segeval` library are right.

**Comment by tomaarsen** [comment]
Gladly! And perhaps, but do note that they return the same values in my quick tests. 

```python
from nltk.metrics import windowdiff
from segeval.window.windowdiff import window_diff
from segeval.format import BoundaryFormat

s1 = "000100000010"
s2 = "000010000100"
s3 = "100000010000"
s4 = "000000000000"
s5 = "111111111111"
print(
    windowdiff(s1, s1, 3),
    window_diff(s1, s1, window_size=3, boundary_format=BoundaryFormat.nltk),
)
print(
    windowdiff(s1, s2, 3),
    window_diff(s1, s2, window_size=3, boundary_format=BoundaryFormat.nltk),
)
print(
    windowdiff(s2, s3, 3),
    window_diff(s2, s3, window_size=3, boundary_format=BoundaryFormat.nltk),
)
print(
    windowdiff(s4, s5, 3),
    window_diff(s4, s5, window_size=3, boundary_format=BoundaryFormat.nltk),
)
```

```
0.0 0
0.3 0.3
0.8 0.8
1.0 1
```
